platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/DefaultConfigurationCacheDegradationController.kt

```diff
@@ -25,21 +25,37 @@ import org.gradle.execution.plan.TaskNode
 import org.gradle.internal.cc.impl.services.DeferredRootBuildGradle
 import org.gradle.internal.service.scopes.Scope.BuildTree
 import org.gradle.internal.service.scopes.ServiceScope
-import java.util.Collections
+import org.gradle.vcs.internal.VcsMappingsStore
 import java.util.concurrent.ConcurrentHashMap
 
 @ServiceScope(BuildTree::class)
 internal class DefaultConfigurationCacheDegradationController(
     private val deferredRootBuildGradle: DeferredRootBuildGradle,
     private val configurationTimeBarrier: ConfigurationTimeBarrier,
+    private val vcsMappingsStore: VcsMappingsStore
 ) : ConfigurationCacheDegradationController {
 
     private val logger = Logging.getLogger(DefaultConfigurationCacheDegradationController::class.java)
     private val tasksDegradationRequests = ConcurrentHashMap<Task, List<Provider<String>>>()
-    private val collectedDegradationReasons = mutableMapOf<Task, List<String>>()
+    private val collectedDegradationReasons = mutableMapOf<DegradationContext, List<String>>()
 
-    val degradationReasons: Map<Task, List<String>>
-        get() = Collections.unmodifiableMap(collectedDegradationReasons)
+    val hasDegradationReasons: Boolean by lazy { collectedDegradationReasons.isNotEmpty() }
```

Comment:
🤔 At my taste this class exposes too many single-used-at-client-site-fields. IMO this class should expose the map of degradation reasons and that's it. All the narrowing down like `hasDegradationReasons`, `hasTaskDegradationReasons` and friends should be declared in the client that need it.

I see how personal this design feeling, so I don't insist on any changes in this regard.

Reply:
The reasoning goes like this: this class is the master of the realm of degradation requests and reasons. The more the features around those concerns are encapsulated here, the better (improves comprehensibility and provides performance improvement opportunities). The lesser those concerns need to be handled elsewhere, the better.  `ConfigurationCacheProblems` is about problems, degradation reasons (or whether there are any) just being one input for how it computes problems.